specs/advanced-data-layouts/prollytree/spec.md

```diff
@@ -0,0 +1,508 @@
+
+[Dolt Implementation Notes — Push And Pull On a Merkle DAG](https://www.dolthub.com/blog/2020-09-09-push-pull-on-a-merkle-dag/)
+
+[mikeal/IPSQL: InterPlanetary SQL](https://github.com/mikeal/IPSQL)
+[mikeal/prolly-trees: Hash consistent search trees.](https://github.com/mikeal/prolly-trees)
+[mikeal/matrika: Next Generation Decentralized Database](https://github.com/mikeal/matrika)
+[mikeal/ipfs-sqlite: SQL on IPFS](https://github.com/mikeal/ipfs-sqlite)
+
+[Merkle Search Trees: Efficient State-Based CRDTs in Open Networks (Scientific Artilcle)](https://hal.inria.fr/hal-02303490)
+  - [simulation](https://gitlab.inria.fr/aauvolat/mst_exp/)
+
+## Summary/Overview
+
+Prolly trees leverage content addressability to create ordered search indexes similar to the common B+ tree structure used for databases, but with the added ability to determenistically merge trees together.
+
+At the highest level, Prolly Trees act as a key value store with the ability to iterate over key ranges which are lexicographically sorted. This property of sorted iteration is important for creating database indexes for a variety of use cases like full text search, sorting of large datasets, and arbitrary search queries.
+
+### Search Tree
+
+The basic structure is that of an ordered search tree: The contained keys are organised such that they can be found (inserted, updated, ...) efficiently by value.
+To efficiently find keys, the tree is traversed top-to-bottom and the non-leaf nodes help navigating/comparing the values efficiently.
+An intermediate node contains several ordered key-address pairs, which link to further nodes (intermediate or leaf) on the next lower level.
+Leaf nodes are identified using an `isLeaf` flag which tells the reader to stop trying to traverse deeper and to treat the values as actual values.
+
+### Chunking
+
+Chunking is the strategy of determining chunk boundaries: Given a list of key-value pairs, it 'decides' which are still inside node A and which already go to the next node B on the same level.
+It depends on the hash/address/node of the items and a 'chunking factor' for tuning the shape of the tree.
+This chunking factor determines the average size of nodes/chunks that a node on a higher level contains.
+This average size in turn controls the shape of the tree (broad/narrow).
+The shape in turn defines the performance of operations on the tree with the following tradeoffs:
+  - larger blocks with more cache invalidations
+  - or smaller blocks with less cache invalidations and more lookups
+
+### Note on Multiple and Single Values and Sets
+
+The described tree can represent a data structure with multiple, single or no values per key. However, given that IPLD Maps (which a Prolly Tree loosely maps to) only allow one value for a key, implementations should merge duplicate keys into a single value. This is also important to have consistent ordering of key value pairs.
```

Comment:
Is there a recommended approach to merging duplicate keys? If I have a database index over a recipe database for "contains meat", my queries will return rows organized by a boolean key, thus being identical for some or even all result rows.

Reply:
I think this is very dependent on the application.

For some applications you can treat the value as a sort of CRDT and merge them, for some you might want a last write wins with some sort of clock, in some cases you might want to have both key value pairs beside each other and iterate over them both when doing a search.

For indexes I usually have the document's primary key in the index key so that they can be unique.

Here's where I did stuff in my hyperbee based DB which is also a key-value store: https://github.com/RangerMauve/hyperbeedeebee/blob/default/index.js#L782